src/chat/chatReducers.js

```diff
@@ -19,139 +16,50 @@ import {
-import chatUpdater from './chatUpdater';
-import { NULL_ARRAY, NULL_OBJECT } from '../nullObjects';
+import { NULL_OBJECT } from '../nullObjects';
 
-const initialState: MessagesState = NULL_OBJECT;
+const initialState: NarrowsState = NULL_OBJECT;
 
 const messageFetchComplete = (
-  state: MessagesState,
+  state: NarrowsState,
   action: MessageFetchCompleteAction,
-): MessagesState => {
+): NarrowsState => {
   const key = JSON.stringify(action.narrow);
-  const messages = state[key] || NULL_ARRAY;
-  const messagesById = groupItemsById(messages);
+  const fetchedMessageIds = action.messages.map(message => message.id);
   const replaceExisting =
     action.anchor === FIRST_UNREAD_ANCHOR || action.anchor === LAST_MESSAGE_ANCHOR;
-  const newMessages = replaceExisting
-    ? action.messages.map(
-        item =>
-          messagesById[item.id]
-            ? isEqual(messagesById[item.id], item)
-              ? messagesById[item.id]
-              : item
-            : item,
-      )
-    : action.messages
-        .filter(x => !messagesById[x.id])
-        .concat(messages)
-        .sort((a, b) => a.id - b.id);
-
   return {
     ...state,
-    [key]: newMessages,
+    [key]: replaceExisting
+      ? fetchedMessageIds
+      : union(state[key], fetchedMessageIds).sort((a, b) => a - b),
   };
 };
 
-const eventReactionAdd = (state: MessagesState, action: EventReactionAddAction): MessagesState =>
-  chatUpdater(state, action.message_id, oldMessage => ({
-    ...oldMessage,
-    reactions: oldMessage.reactions.concat({
-      emoji_name: action.emoji_name,
-      user: action.user,
-    }),
-  }));
-
-const eventReactionRemove = (
-  state: MessagesState,
-  action: EventReactionRemoveAction,
-): MessagesState =>
-  chatUpdater(state, action.message_id, oldMessage => ({
-    ...oldMessage,
-    reactions: oldMessage.reactions.filter(
-      x => !(x.emoji_name === action.emoji_name && x.user.email === action.user.email),
-    ),
-  }));
-
-const eventNewMessage = (state: MessagesState, action: EventNewMessageAction): MessagesState => {
-  let stateChange = false;
-  const newState = Object.keys(state).reduce((msg, key) => {
+const eventNewMessage = (state: NarrowsState, action: EventNewMessageAction): NarrowsState =>
+  Object.keys(state).reduce((msg, key) => {
     const isInNarrow = isMessageInNarrow(action.message, JSON.parse(key), action.ownEmail);
     const isCaughtUp = action.caughtUp[key] && action.caughtUp[key].newer;
-    const messageDoesNotExist =
-      state[key].find(item => action.message.id === item.id) === undefined;
+    const messageDoesNotExist = state[key].find(id => action.message.id === id) === undefined;
 
     if (isInNarrow && isCaughtUp && messageDoesNotExist) {
-      stateChange = true;
-      msg[key] = [...state[key], action.message];
+      msg[key] = [...state[key], action.message.id];
     } else {
       msg[key] = state[key];
     }
-
     return msg;
   }, {});
 
-  return stateChange ? newState : state;
```

Comment:
This `stateChange` logic kind of looks like it's doing nothing, but it's actually providing an optimization of a kind that's very important in the React architecture: when the new state is substantively unchanged from the old, it causes it to actually be the same state object, equal under `===`.

That's important because it allows downstream users of this part of the state -- notably React in handling components that get it as a prop, and Reselect in handling `createSelector` selectors where it's an argument -- to see that it's unchanged and cut off any downstream processing right here.

The exact structure of this code feels a little bit clumsy to me, and I'd welcome something to tighten it. (Separately from the main commit here, though it'd be fine to be in the same branch if that's convenient.) But we should keep the optimization. :-)

Reply:
As a bit of commit-style feedback, this simplification would have been best as a separate, preparatory commit. That way (in a world where the simplification is correct) you'd get to take full advantage of it in simplifying the code to be edited in the main commit, and at the same time it has its own commit message and can be read and understood separately. And keeps the complexity of the main commit down by not adding to the changes that are part of that commit.

Reply:
So my original plan was to ignore all optimizations initially and wait for someone with an old phone to benchmark the app, and then decide on whether to put optimizations back in. But I'll take your word for this one being critical :)

Reply:
Also it just sorta felt nice to have the core-core event `eventNewMessage` be simply

```js
const eventNewMessage = (state: MessagesState, action: EventNewMessageAction): MessagesState => ({
  ...state,
  [action.message.id]: action.message,
});
```

:P

Reply:
Hmm, thinking about it a bit more, I'm not sure we should keep this optimization: For one, I don't know how to implement it: This commit splits chatReducers into chatReducers and messageReducers, and messageReducers only has access to the new `messages` state. It is therefore unaware of the `narrows` state that is needed for this optimization (`narrows` corresponds to `chatReducers`). 

For two, I'm not sure we wanna do the optimization here: If we do proper optimization by carefully detecting unchanged arrays of messages (as discussed here https://chat.zulip.org/#narrow/stream/243-mobile-team/subject/message.20state/near/618189), this shouldn't be necessary. Also, we need to do the unchanged-arrays-of-messages-optimization in both cases. 

So I'll wait for your input on this.

Reply:
But here on the `chat` aka `narrows` side, I think we still need essentially the same optimization we have in the old code, for the same reason: most importantly, there will be a lot of messages coming in where we don't have that narrow, or we have it somewhere in the past but aren't caught up. And I think it should be easy to keep with basically the same logic.

Reply:
... Hmm, now I see what you mean by your remark about the split, though: there's a sort of emergent property the existing design has, which is that if a message comes in that doesn't fall into any narrow we're tracking, then we won't store the message itself. And that does become tricky to preserve.

Is that optimization essential? Let's see...
* On a busy day chat.zulip.org (an especially large and busy org) might see 1000 messages. Say 100/hr.
* At that rate, the messages we collect over even a few hours aren't a big deal. And nobody will use the app continuously, with no breaks longer than 10 minutes (the event queue timeout), for more than a few hours at a time.
* So we don't need it right now.
* In the future, we intend to use a longer-lived event queue for some events; this is mainly for non-messages, but could include some messages (PMs? pinned streams?)
* Even then, I think we'll only do that for messages that we want to keep in any case. And if not... we can cross that bridge when we come to it.

So for *that* optimization, I agree -- let's just go without it.

We should keep the ones that are easy to keep, though. :smile: Including the one on new messages in `chatReducers` that's implemented with this `stateChange` logic -- which is a bit clumsy but OK, and which we can basically keep unchanged.

Reply:
So after looking a bit more into this, I found out how it'd be possible to preseve the isInNarro optimization: We'd need to pass the `narrows` state as part of `EventNewMessageAction`. This is in fact already done for another subtree of the state, `caughtUp` (which enables the caughtUp optimization.

So this optimization is doable without too much work, but it should go into a new commit. I'll not implement that commit now, but it should be doable as a follow-up. Also, since the isCaughtUp optimization is tightly coupled to the isInNarrow optimization, I'll skip the isCaughtUp one for now as well (at least for messageReducers.js).